### ISSUE TITLE
Collection columns of UDT values

### ISSUE THREAD

**Issue by prettynatty** [outworkers/phantom#157]
CQL lets define columns typed as collections of UDTs.
Is there a way to use it in phantom?

**Comment by MartinHerren** [comment]
I'm adding myself to the list of those waiting for UDT support :-)

Still planned to have an initial version out in February ?

In the meantime i tried to play with what's already available, inspiring myself from the test file https://github.com/websudos/phantom/blob/742bca90ce44b467644a757e488d8e86c684c0dc/phantom-udt/src/test/scala/com/websudos/phantom/udt/Tables.scala but that results in a "java.lang.CloneNotSupportedException: TestRecords$t$", where t is my UDTColumn object, when i try to do a select. I tried naively (i'm still new to Scala) to mix in the trait Clonable on the object but that results in a java.NullPointerException.

Is it me that is doing something wrong or is it not supposed to work yet ? Tested with 1.5.0 and 1.5.1. Is there already more in HEAD ?

The table creation fails as well on a C* InvalidQueryException as it tries to use a non-frozen UDT, but that doesn't matter to me yet as i only need to read an already created and populated table.

But thanks already for the great work done on Phantom !

**Comment by alexflav23** [comment]
@MartinHerren  @prettynatty @elmalto A small update on UDTs, which unfortunately didn't make February as we prioritised a new QueryBuilder release ahead of UDT support to make it far easier to add idiomatic UDT support to start with.

They have now however reached the top of the priority list, and we are hoping they will be out very soon, within the next 2 - 3 weeks. They will be included in the grand launch of phantom 2.0.0, which is a very large re-writing of critical bits such as Query serialization.

**Comment by alexflav23** [comment]
@MartinHerren @prettynatty @elmalto Another update on this. The new Querybuilder is not out in 1.8.0 and UDT support is officially top priority, so do except to see at least a prototype level working version very shortly, hopefully not longer than 2 weeks.

**Comment by coreyauger** [comment]
Just wanted to check if there are any updates on this.  Thanks.

**Comment by alexflav23** [comment]
Hi @coreyauger,

We have an internal version of UDTs but we have yet to put in the effort to stabilise things, mainly as we have a lot of other high priority things on the TODO list. We will update this topic when we have further news.

Regards,
Flavian

**Comment by shayangz** [comment]
any update on this front?

**Comment by andyzeli** [comment]
Kudos to Websudos on this phenomenal driver. however we really are in need of UDT support too...

**Comment by Ciaran0** [comment]
Any update on this in 2016?

**Comment by alexflav23** [comment]
Hi @andyzeli , @Ciaran0,

Yes, it's going to be available in phantom 2.0 starting with end of April/May. We are working very hard to bring in a lot of new functionality to phantom.

Regards.

**Comment by favalos** [comment]
Hi,

What is the status of this? Was it released already?

Regards,

**Comment by alexflav23** [comment]
Hi @favalos,

There's an internal distribution we are still testing, we have UDT support for everything except collection columns which present some peculiarities because of Scala 2.10 macro bugs in the compiler/macro framework itself.

If you want to try out a beta release, send me an email and I'll give you beta access, but obviously you may need to help us with feedback and bug reports as needed.

Regards.

**Comment by alexflav23** [comment]
Hi @favalos @prettynatty @Ciaran0 @andyzeli @coreyauger @elmalto @shayangz,

One of the most anticipated and long winded deliveries is finally making its way to phantom. UDT support will be available from `0.4.0` as part of phantom-pro, which will be our subscription based product to include spark support, automated migrations, a tool to manage migrations and much more.

Now as early adopters you can benefit from a heavily reduced license for the first year, at a nominal fee of a few dollars a month (probably $5), just so we can test our integration. All proceeds of any commercial stuff will go directly back into open source work and building up more and more projects.

UDT support looks like this, so it's pretty much invisible. This uses a very complex backing mechanism and recursive implicits among other things to allow you to nest things however you want and so forth.

If you would like to be in the early beta release, please shout or send me an email directly, flavian@outworkers.com, we are days away from all this being available for use, and its extensively tested(as much as a beta can be), so you should have no trouble with some production fun straight away.

```scala
@Udt case class CustomType(
  id: UUID,
  name: String,
  email: String
)

case class Record(id: UUID, list: List[String], custom: CustomType)

class MyTable extends CassandraTable[MyTable, Record] {
  object id extends UUIDColumn(this) with PartitionKey[UUID]
  object list extends ListColumn[String](this)
  object custom extends UDTColumn[MyTable, Record, CustomType]

  def fromRow(row: Row): Record = Record(id(row), list(row), custom(row))
}
```
